### PR TITLE
Memory regression test [blocks: #2035]

### PR DESCRIPTION
**Pull request by tautschnig**
This is a first proof of concept that should help spotting differences in heap memory usage. CPU performance benchmarking does not necessarily provide sufficient insight as discussed in #1855 and #1860.

### CODE REVIEW

File: scripts/memory-test/memory.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+                snapshot(
+                    reference_data['snapshots'][rs],
+                    rs == r_peak_index))
+
+    data_seq = []
+    for rd in range(0, len(data['snapshots'])):
+        data_seq.append(
+                snapshot(data['snapshots'][rd], rd == peak_index))
+
+    diff = color_diff(
```
Comment (chrisr-diffblue) [inline scripts/memory-test/memory.py]: While colour diff output is pretty and a neat feature, could we make it fallback gracefully to non-colourised output when, e.g. the output terminal doesn't support colour (rare these days) or the output is piped to a file?

File: .travis.yml [hunk 2]
```diff
@@ -182,11 +182,10 @@ jobs:
         - CCACHE_CPP2=yes
       script: echo "Not running any tests for a debug build."
 
-    # cmake build using g++-5
+    # cmake build using g++-5 + memory regression test
     - stage: Test different OS/CXX/Flags
       os: linux
       compiler: gcc
-      cache: ccache
       env:
```
Comment (chrisr-diffblue) [inline .travis.yml]: What is the reason for disabling ccache here? You changed this config to be a Debug build, but you can still use ccache with that if you increase the ccache size to 2G just for this configuration.

File: .travis.yml [hunk 3]
```diff
@@ -195,14 +194,22 @@ jobs:
...
           packages:
             - g++-5
+            - valgrind
       before_install:
         - mkdir bin ; ln -s /usr/bin/gcc-5 bin/gcc
       install:
-        - ccache -z
-        - ccache --max-size=1G
-        - cmake -H. -Bbuild '-DCMAKE_BUILD_TYPE=Release'  '-DCMAKE_CXX_COMPILER=/usr/bin/g++-5'
+        - cmake -H. -Bbuild '-DCMAKE_BUILD_TYPE=Debug'  '-DCMAKE_CXX_COMPILER=/usr/bin/g++-5'
```
Comment (chrisr-diffblue) [inline .travis.yml]: Do you want/need Debug here, or would RelWithDebInfo be a closer representative of the memory performance of a release build?

File: .travis.yml (hunk 2 above)
Comment (tautschnig) [reply .travis.yml]: Is that ok to do? Surely happy to do that, I was actually hoping for your input on that ;-)

File: .travis.yml (hunk 3 above)
Comment (tautschnig) [reply .travis.yml]: Probably true - will use that!

File: .travis.yml (hunk 2 above)
Comment (chrisr-diffblue) [reply .travis.yml]: Yeah, I'd much rather have ccache enabled and bump it's size just for this config than not have it enabled (it'll make saving the cache back a bit slower, but you'll more than win that back with the build time improvements :-) )

File: scripts/memory-test/memory.py (hunk 1 above)
Comment (tautschnig) [reply scripts/memory-test/memory.py]: Done, I think. (Colorama knows how to do this, and I've now hacked that into the `ColorFallback` code as well.)

**Comment by tautschnig** [comment]
This appears to work reliably now. Reviews would be appreciated.

**Review by chrisr-diffblue — APPROVED** [review]
Looks good to me!

**Comment by tautschnig** [comment]
Just one heads-up: it seems this regression test is now stable when run on an unmodified code-base. The memory-profile matching is fuzzy, but might not yet be sufficiently fuzzy to accept all negligible differences. I will have to work on this once it is merged and spuriously failing PRs arise.

**Comment by forejtv** [comment]
@tautschnig I am not too excited about having a subdirectory with an alternative license. Where does the script come from? Any chance we could use it under the standard CBMC license?

**Comment by tautschnig** [comment]
@forejtv Indeed my prior setup was misleading as it's only the externally sourced msparser.py that is meant to be under the MIT license; I have moved the license information into the file. See commit message for details. Does that seem acceptable?

**Comment by chrisr-diffblue** [comment]
I haven't gone to look what other licenses we have inside the CBMC repo, but if we're going to have mixed license like this, is there an argument for putting all the files that are separately licensed under their own directory hierarchy - e.g a 'contrib' or 'external' directory at the top level of the CBMC repo? Or treat these files like mini-sat, where they get downloaded rather than imported into the repo? I'm not personally hugely concerned either way, but it would make things clearer if all the 'non CBMC license' stuff was more obviously separated.

**Comment by tautschnig** [comment]
As far as I'm aware we have got:
* scripts/cpplint.py: Copyright Google Inc; BSD 3-clause license
* src/big-int/*: Copyright Dirk Zoller; public domain
* src/miniz/*: Copyright RAD Game Tools, Valve Software, Rich Geldreich, Tenacious Software LLC; MIT license
* src/nonstd/optional.hpp: Copyright Martin Moene; MIT license
* src/util/sorted_vector.h: Copyright Martin Holzherr; MIT-like license (#2019 proposes to remove this file)
* unit/testing-utils/catch.hpp: Copyright Two Blue Cubes Ltd; Boost Software License (almost the same as MIT)

**Comment by tautschnig** [comment]
Back to reviewers: not sure what the status on licenses is, otherwise this has now seen the following improvements: 1) The memory regression test is non-fatal to make sure its somewhat experimental nature doesn't cause inadvertent blocking. 2) The diff script now is fuzzy in its matching so that the the varying number of snapshots that massif generates does not cause havoc each time. 3) I have experimented with Google perftools and heaptrack and have neither found either of them more deterministic nor more insightful; on the contrary, massif still produces the most reliable memory profiles.

File: .travis.yml [hunk 4]
```diff
@@ -199,11 +199,44 @@ jobs:
         - mkdir bin ; ln -s /usr/bin/gcc-5 bin/gcc
       install:
         - ccache -z
-        - ccache --max-size=1G
-        - cmake -H. -Bbuild '-DCMAKE_BUILD_TYPE=Release'  '-DCMAKE_CXX_COMPILER=/usr/bin/g++-5'
+        - ccache --max-size=2G
+        - cmake -H. -Bbuild '-DCMAKE_BUILD_TYPE=RelWithDebInfo'  '-DCMAKE_CXX_COMPILER=/usr/bin/g++-5'
```
Comment (NathanJPhillips) [inline .travis.yml:203]: Have these two changes accidentally escaped here from other work?

**Review by NathanJPhillips — COMMENTED** [review]
Don't know enough about memory profiling, Python or Yaml to contribute a useful review however I approve of the efforts to introduce memory profiling to CBMC.

File: .travis.yml (hunk 4 above)
Comment (tautschnig) [reply .travis.yml:203]: No, they are intentional; the commit message tries to call them out, but it might not be sufficiently clear?

Comment (chrisr-diffblue) [reply .travis.yml:203]: I can confirm that this was an intentional change that @tautschnig and I 
discussed in an earlier part of this PR.

**Review by martin-cs — APPROVED** [review]
I'm not going to comment on the licensing issues but massif is my tool of choice for heap profiling CPROVER and having a regular, automated run of it would be A Good Thing.

**Review by peterschrammel — APPROVED** [review]
Putting the MIT-licensed test utility scripts into a separate directory is ok.

**Comment by TGWDB** [comment]
Closing this this is very out of date and apparently untouched. The blocker appears to no longer be true. Please reopen and redo for new CI system if you believe this is erroneous.

**Pull request closed without merging**